lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+// returns a sum of CP value between every pair of fields on the block
+double FieldReorderAnalyzer::calculateWCPWithinCacheBlock(
+    std::vector<FieldNumType> *CacheBlock) const {
+  double WCP = 0;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Calculate WCP within a cache block: ");
+  for (unsigned i = 0; i < CacheBlock->size(); i++) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                    dbgs() << "F" << (*CacheBlock)[i] + 1 << " ");
+    for (unsigned j = i + 1; j < CacheBlock->size(); j++)
+      if ((*CacheBlock)[i] < (*CacheBlock)[j])
+        WCP += CloseProximityRelations[(*CacheBlock)[i]][(*CacheBlock)[j]];
+      else
+        WCP += CloseProximityRelations[(*CacheBlock)[j]][(*CacheBlock)[i]];
+  }
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "produces WCP: " << DEBUG_PRINT_COUNT(WCP) << "\n");
+  return WCP;
+}
+
+// Calculate a total WCP of all fields that are already in NewOrder. Return
+// the total WCP. This function is used to measure if the current NewOrder
+// is a good ordering
+double FieldReorderAnalyzer::getWCP() const {
+  unsigned CurrentCacheBlockSize = 0;
+  std::vector<FieldNumType> CurrentCacheBlock;
+  double WCP = 0;
+  for (auto &FieldNum : NewOrder) {
+    auto Size = FieldSizes[FieldNum];
+    DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER, dbgs() << "Field " << FieldNum + 1
+                                               << " size is " << Size << "\n");
+    if (!StructureType->getElementType(FieldNum)->isAggregateType() &&
+        CurrentCacheBlockSize % Size) {
+      // Alignment needed, need to add padding. Ignore alignment for aggregated
+      // type (array and struct)
+      DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                      dbgs() << "Need to add "
+                             << (CurrentCacheBlockSize / Size + 1) * Size -
+                                    CurrentCacheBlockSize
+                             << " Bytes of padding\n");
+      CurrentCacheBlockSize = (CurrentCacheBlockSize / Size + 1) * Size;
+      assert(CurrentCacheBlockSize <= CacheBlockSize);
+    }
+    if (CurrentCacheBlockSize == CacheBlockSize) {
+      DEBUG_WITH_TYPE(
+          DEBUG_TYPE_REORDER,
+          dbgs()
+              << "Need to put into a new cache block calculate current one\n");
+      WCP += calculateWCPWithinCacheBlock(&CurrentCacheBlock);
+      CurrentCacheBlockSize = 0;
+      CurrentCacheBlock.clear();
+    } else {
+      CurrentCacheBlockSize += Size;
+      CurrentCacheBlock.push_back(FieldNum);
+      if (CurrentCacheBlockSize >= CacheBlockSize) {
+        DEBUG_WITH_TYPE(
+            DEBUG_TYPE_REORDER,
+            dbgs() << "Current cache block cannot fit remaining field\n");
+        WCP += calculateWCPWithinCacheBlock(&CurrentCacheBlock);
+        CurrentCacheBlockSize %= CacheBlockSize;
+        CurrentCacheBlock.clear();
+        if (CurrentCacheBlockSize > 0)
```

Comment:
I reason I put them in the same checking because whether adding the field exceeds limit or just fills the current cache block, we always want to push the current block and calculate the WCP. The only difference is after the push and calculation, exceeding case needs to push the current field to new cache block while the equal case doesn't need to. If we split them into two cases, there would be some repeated code and just one line of difference.
I've added more comments in this part so hopefully it's not that confusing. Do you agree?